#### This script is a standardized pipeline for RNASeq analysis 
#### Author: Abhishek Das
#### Description: This pipeline takes count files generated by HTSeq-count or 
#### featureCount and calculate correlation and PCA analysis between samples.
#### It also generates correlation, PCA and MA plots.
#### Date: 2-11-2021



################### Loading libraries and Data for analysis ####################

library("stringi")
library("DESeq2")
library("apeglm")

directory = "C:/data/reads/fallback/A2780_PANPOTE_RNA_SEQ/A2780_PANPOTE_RNA_SEQ/CountFiles"  #Path to sample files
setwd(directory)
sampleFiles = list.files(directory)
sampleTable = read.table("sampleFiles.txt", header = TRUE, sep = "\t") #Change sample file name
sampleCondition = sampleTable$condition

directory= "C:/data/reads/fallback/A2780_PANPOTE_RNA_SEQ/A2780_PANPOTE_RNA_SEQ/CountFiles/Counts" #Path to count files

dds <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,
                                  directory = directory,
                                  design= ~ condition)
dds <- DESeq(dds)
res <- results(dds)

############### Effects of transformations on the variance #####################
#### this gives log2(n + 1)

#### Data transformations and visualization
#### Count data transformations

vsd <- vst(dds, blind=FALSE)

ntd <- normTransform(dds)

library(vsn)

meanSdPlot(assay(ntd))

meanSdPlot(assay(vsd))

#### Heatmap of the sample-to-sample distances

library("pheatmap")
library("RColorBrewer")

sampleDists <- dist(t(assay(vsd)))

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$type, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

plotPCA(vsd)

#### Approach to count outliers
par(mar=c(8,5,2,2))
boxplot(log10(assays(dds)[["cooks"]]), range=0, las=2, cex.axis=0.6)

#### Dispersion plot and fitting alternatives
plotDispEsts(dds)


################ Exploring and exporting results and MA plot ###################
#### Log fold change shrinkage for visualization and ranking

resultsNames(dds)
resLFC <- lfcShrink(dds, coef=resultsNames(dds)[2], type="apeglm")
resLFC


#### Other Interesting MA Plot
# load library
library(ggrepel)

# Creating a dataframe
de <- data.frame(
  gene_symbol = row.names(res), 
  log2FoldChange= res$log2FoldChange,
  pvalue  = res$pvalue)

ggplot(data= de, aes(x=log2FoldChange, y=pvalue)) + geom_point()

# add a column of NAs
de$diffexpressed = "NO"

#if log2Foldchange > 1.5 and pvalue < 0.05, set as "UP" 
de$diffexpressed[de$log2FoldChange > 1.5 & de$pvalue < 0.05] <- "UP"

# if log2Foldchange < -0.6 and pvalue < 0.05, set as "DOWN"
de$diffexpressed[de$log2FoldChange < -1.5 & de$pvalue < 0.05] <- "DOWN"

p <- ggplot(data=de, aes(x=log2FoldChange, y=-log10(pvalue), col=diffexpressed)) + geom_point() + theme_minimal()

p2 <- p + geom_vline(xintercept=c(-1.5, 1.5), col="red") + geom_hline(yintercept=-log10(0.05), col="red")

p3 <- p2 + scale_color_manual(values=c("blue", "black", "red"))

de$delabel <- NA
de$delabel[de$diffexpressed != "NO"] <- de$gene_symbol[de$diffexpressed != "NO"]

ggplot(data=de, aes(x=log2FoldChange, y=-log10(pvalue), col=diffexpressed, label=delabel)) + 
  geom_point() + 
  theme_minimal() +
  geom_text()

# plot adding up all layers we have seen so far
ggplot(data=de, aes(x=log2FoldChange, y=-log10(pvalue), col=diffexpressed, label=delabel)) +
  geom_point() + 
  theme_minimal() +
  geom_text_repel() +
  scale_color_manual(values=c("blue", "black", "red")) +
  geom_vline(xintercept=c(-0.6, 0.6), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")


# Writing the differntial Exprssion data 

plotMA(res, ylim=c(-1.5,1.5))
plotMA(resLFC, ylim=c(-1.5,1.5))
resOrdered <- res[order(res$pvalue),]
summary(res)
sum(res$padj < 0.1, na.rm=TRUE)
res05 <- results(dds, alpha=0.05)
summary(res05)

resdata = merge(as.data.frame(res), as.data.frame(counts(dds, normal=TRUE)), by="row.names", sort=FALSE)

write.table(resdata,paste(resultsNames(dds)[2],".xls"), row.names = T, col.names = T) #Change output file name


###Heapmap for top10 upregulated and downregulated genes

sig <- rownames(resdata)[which(resdata$pvalue < .05 & !is.na(resdata$padj))]
resSig=resdata[sig,]
HMdat =as.data.frame(resSig[order(resSig$log2FoldChange, decreasing=T)[1:10],])
HMdat=rbind(HMdat,resSig[order(resSig$log2FoldChange, decreasing=F)[1:10],])
HMdat[order(HMdat$log2FoldChange),]
rownames(HMdat)=HMdat$Row.names
#resSig
pheatmap(scale(as.matrix(HMdat[,8:ncol(HMdat)])), 
         cluster_rows=T, show_rownames=TRUE,
         cluster_cols=T)

library(dplyr)

downRegulated = filter(resdata,resdata$log2FoldChange < -1.5 & resdata$pvalue < 0.05)[,1]

write.table(downRegulated,"downregulated", row.names = F, col.names = T)

upRegulated = filter(resdata,resdata$log2FoldChange > 1.5 & resdata$pvalue < 0.05)[,1]

write.table(upRegulated,"upregulated", row.names = F, col.names = T)

################################################################################


library(biomaRt)
ensembl = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
gene2genomeEx <- getBM(values = upRegulated,
                       filters = "hgnc_symbol",
                       mart = ensembl,
                       attributes = c("ensembl_gene_id", "entrezgene_id",
                                      "hgnc_symbol", "external_gene_name",
                                      "description", "chromosome_name",
                                      "strand"))

############################## GO Analysis ##############################
#### cluster profiler step:
library(org.Hs.eg.db)
library(clusterProfiler)

mylistt<- gene2genomeEx$entrezgene_id
na.omit(mylistt)

ggo = groupGO(gene=as.character(gene2genomeEx$entrezgene_id), OrgDb='org.Hs.eg.db', ont = "CC", level = 4, readable = FALSE)
barplot(ggo, drop=TRUE, showCategory=15)

ggo = groupGO(gene=as.character(gene2genomeEx$entrezgene_id), OrgDb='org.Hs.eg.db', ont = "MF", level = 4, readable = FALSE)
barplot(ggo, drop=TRUE, showCategory=15)

ggo = groupGO(gene=as.character(gene2genomeEx$entrezgene_id), OrgDb='org.Hs.eg.db', ont = "BP", level = 4, readable = FALSE)
barplot(ggo, drop=TRUE, showCategory=15)
